src/Features/LanguageServer/Protocol/LspServices/LspServices.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition.Hosting;
+using System.Linq;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer
+{
+    internal class LspServices : IDisposable
+    {
+        public ImmutableDictionary<Type, Lazy<ILspService, LspServiceMetadataView>> LazyLspServices { get; }
+
+        public LspServices(ExportProvider exportProvider, WellKnownLspServerKinds serverKind, ImmutableArray<Lazy<ILspService, LspServiceMetadataView>> baseServices)
+        {
+            // The act of calling get exports on an ILspService will instantiate new instances as long as they are not exported with the Shared attribute.
+            // This is the key codepath that ensures each LSP server gets their own set of LSP services on initialization.
+            var services = exportProvider.GetExports<ILspService, LspServiceMetadataView>(serverKind.GetContractName());
```

Comment:
@sharwell I wanted to get your MEF expertise on this to see if this was the best way to do this.  Basically what I want is to be able to create new instances of the `ILspService` for each LSP server instance (could be multiple running at one time), dispose of them on shutdown, and have them be recreated properly if the server shutsdown / restarts.

The best way I could come up with is making all exports for `ILspService` not shared and manually retrieve them on server creation here to force mef to create new ones.  

However this will require me to suppress `#RS0023 // Parts exported with MEFv2 must be marked with 'SharedAttribute'` as these cannot be shared instances.  I think I can write a diagnostic suppressor for that?  But wanted to check with you first if you had a better approach.

Reply:
Rather than export `ILspService`, you can export `ILspServiceFactory`. This allows the factory to be static, and also allows the factory to cache the service instance if it wants to. See `IWorkspaceServiceFactory` for another example.

Reply:
basically did this - there is some extra boiler plate with this approach (e.g. https://github.com/dotnet/roslyn/pull/61266/files#diff-e49357a9ef274059c786a87abe0c688392ac8101a18dc41327d1aaa347c17243R20 ) but the majority are stateless and so don't need to actually use a factory.  

To indicate which is which I have two attributes - one for exporting stateless LSP services, and one for exporting lsp service factories